### ISSUE TITLE
Telemtry v2 does not report metrics for mirror traffic at source, but does destination

### ISSUE THREAD

**Issue by Stono** [istio/istio#23496]
**Bug description**
Hey, 
At the moment if you enable traffic mirroring, you do not get any `reporter=source` metrics for the mirror traffic, but you will get `reporter=destination` metrics at the other end.

This leads to confusion when looking at metrics because the destination is reporting very different numbers.

**Expected behavior**
My gut is that this should be consistently reported on both sides of the request.  Either in neither, or in both

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.5.2

**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by mandarjog** [comment]
@douglas-reid related to https://github.com/istio/istio/issues/11475

**Comment by douglas-reid** [comment]
yeah, we should start a doc on mirroring and what we would expect that to look like. do we need some way to indicate the traffic was mirrored ?

**Comment by douglas-reid** [comment]
see also: #11093

**Comment by esnible** [comment]
@mandarjog is this still 1.7?

**Comment by douglas-reid** [comment]
there is no work that I know of that was done for this in 1.7

**Comment by Stono** [comment]
Seeing as there are several things like this since telemetryv2, is it correct to continue to mark things like “client and server telemetry reporting” and “traffic control” as “Stable” on the features https://istio.io/latest/about/feature-stages ?

Stable being “Dependable, production hardened”

**Comment by douglas-reid** [comment]
@Stono i understand your frustration. there are definitely gaps that remain between v1 and v2 that we need to be clear about.

For this particular issue, however, I'm not sure this represents a regression. I don't believe we've ever had source telemetry working for mirrored traffic. Issues #11475 and #11093 (linked above) pre-date the switch to proxy-based telemetry.

This is something that definitely needs to be fixed and _should_ be supported, however. We've had a number of proposals for workarounds and potential fixes throughout the evolution of the project, but they've all been rejected or abandoned, unfortunately.

It is too late to get this work into 1.7, but we can prioritize it for 1.8.

/cc @mandarjog

**Comment by kyessenov** [comment]
I don't think there's anything for mirrored traffic at the source. All istio features (authz, networking, telemetry) do not apply for mirrored traffic at the source. I think we managed to fix the trace propagation but that's about it.

**Comment by Stono** [comment]
Similar problem but more serious manifestation is when you get DC errors (source service disconnecting, for example - because it's hit a read timeout) you don't get metric labels either: https://github.com/istio/istio/issues/25185

**Comment by Stono** [comment]
Also associated https://github.com/istio/istio/issues/21923

**Comment by douglas-reid** [comment]
not stale

**Comment by Stono** [comment]
@douglas-reid @howardjohn please can we reopen this?  It's still a problem and had me scratching my head today until I found my own issue! :D 

There are no metrics in the `reporter=source` for traffic mirror'd data.

Or are we saying this _won't_ be fixed?

**Comment by douglas-reid** [comment]
> Or are we saying this _won't_ be fixed?

This work has never really been prioritized. It probably should be looked at in more depth. I've long since lost the full context. It _feels_ like a more holistic approach to handling mirrored traffic is warranted (not just a telemetry issue). And the work likely involves some proxy-side dev too. 

@kyessenov may remember more here, as I believe he investigated this in more detail previously.

**Comment by kyessenov** [comment]
Yea, the reason is pretty simple. Envoy mirroring is fire-and-forget, so we can't attach any of our custom filters on that datapath. I think all we need is a special header injected.

**Comment by kyessenov** [comment]
Let me rephrase - mirroring initiates an async call through an upstream cluster bypassing any more downstream processing. Because we attach our logic to downstreams, that does not apply to mirrored traffic. So there are several ways to improve it:
* extend async client with ability to send MX header;
* extend upstream clusters to send MX header (WIP upstream to add upstream HTTP filters);
* route through another downstream for mirrored traffic (this can be done with a reserved listener that then forwards upstream).

**Comment by Stono** [comment]
Just adding another example use case for this today.

We wanted to compare the performance of a next version of a remote api and immediately traffic mirroring came to mind as a quick and easy way to duplicate everything.

But then we remembered that we won't see the metrics for the traffic mirror, so we weren't able to use it to compare the response times for the old and new systems :-(

**Comment by kyessenov** [comment]
There's a new demand for improving mirrored requests: https://github.com/envoyproxy/envoy/issues/21684. We can already set-up move metadata exchange to the upstream filters which solves the issue of metadata propagation.

**Comment by Stono** [comment]
@kyessenov that issue was closed as stale, and also matts comments indicate another larger piece of work that _might_ help fix it.

Is this something that can't be fixed in Istio, only envoy?  Not having l7 metrics for mirroring is such a limitation :-(
